File: clients/src/main/java/org/apache/kafka/common/Endpoint.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.common;

import org.apache.kafka.common.annotation.InterfaceStability;
import org.apache.kafka.common.security.auth.SecurityProtocol;

import java.util.Objects;
import java.util.Optional;

/**
 * 表示一个Broker端点。
 * 端点包含了Broker的监听器名称、安全协议、主机名和端口等信息，
 * 用于客户端与Broker建立连接。
 */
@InterfaceStability.Evolving
public class Endpoint {

    // 监听器名称，用于标识特定的监听器配置
    private final String listenerName;
    // 安全协议类型，如PLAINTEXT、SSL、SASL等
    private final SecurityProtocol securityProtocol;
    // Broker的主机名或IP地址
    private final String host;
    // Broker监听的端口号
    private final int port;

    /**
     * 创建一个新的端点实例
     * @param listenerName 监听器名称
     * @param securityProtocol 安全协议
     * @param host 主机名
     * @param port 端口号
     */
    public Endpoint(String listenerName, SecurityProtocol securityProtocol, String host, int port) {
        this.listenerName = listenerName;
        this.securityProtocol = securityProtocol;
        this.host = host;
        this.port = port;
    }

    /**
     * 返回端点的监听器名称。
     * 对于提供给broker插件的端点，该值不为空；
     * 但在客户端使用时可能为空。
     */
    public Optional<String> listenerName() {
        return Optional.ofNullable(listenerName);
    }

    /**
     * 返回端点使用的安全协议
     */
    public SecurityProtocol securityProtocol() {
        return securityProtocol;
    }

    /**
     * 返回端点的已配置主机名
     */
    public String host() {
        return host;
    }

    /**
     * 返回监听器绑定的端口号
     */
    public int port() {
        return port;
    }

    /**
     * 比较两个端点是否相等
     * 当所有字段(监听器名称、安全协议、主机名、端口)都相等时返回true
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Endpoint)) {
            return false;
        }

        Endpoint that = (Endpoint) o;
        return Objects.equals(this.listenerName, that.listenerName) &&
            Objects.equals(this.securityProtocol, that.securityProtocol) &&
            Objects.equals(this.host, that.host) &&
            this.port == that.port;

    }

    /**
     * 生成端点的哈希码
     * 使用所有字段计算哈希值
     */
    @Override
    public int hashCode() {
        return Objects.hash(listenerName, securityProtocol, host, port);
    }

    /**
     * 返回端点的字符串表示
     * 包含所有字段的值
     */
    @Override
    public String toString() {
        return "Endpoint(" +
            "listenerName='" + listenerName + "'" +
            ", securityProtocol=" + securityProtocol +
            ", host='" + host + "'" +
            ", port=" + port +
            ")";
    }
}

File: core/src/main/scala/kafka/cluster/EndPoint.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.cluster

import org.apache.kafka.common.{KafkaException, Endpoint => JEndpoint}
import org.apache.kafka.common.network.ListenerName
import org.apache.kafka.common.security.auth.SecurityProtocol
import org.apache.kafka.common.utils.Utils

import java.util.Locale

/**
 * EndPoint对象提供了用于处理Kafka broker网络端点相关的工具方法
 */
object EndPoint {
  /**
   * 从连接字符串中解析监听器名称
   * 
   * @param connectionString 格式为"listenerName://host:port"的连接字符串
   * @return 提取并转换为大写的监听器名称
   * @throws KafkaException 当无法从连接字符串中解析出监听器名称时抛出异常
   */
  def parseListenerName(connectionString: String): String = {
    // 查找第一个冒号的位置，用于分隔监听器名称和主机端口信息
    val firstColon = connectionString.indexOf(':')
    if (firstColon < 0) {
      throw new KafkaException(s"Unable to parse a listener name from $connectionString")
    }
    // 提取并转换为大写的监听器名称
    connectionString.substring(0, firstColon).toUpperCase(Locale.ROOT)
  }

  /**
   * 将Java版本的Endpoint转换为Scala版本的EndPoint
   * 
   * @param endpoint Java版本的端点对象
   * @return 转换后的Scala版本EndPoint对象
   */
  def fromJava(endpoint: JEndpoint): EndPoint =
    new EndPoint(endpoint.host(),
      endpoint.port(),
      new ListenerName(endpoint.listenerName().get()),
      endpoint.securityProtocol())
}

/**
 * Kafka broker的网络端点定义，用于表示broker的监听地址
 * 
 * 每个EndPoint包含了以下核心信息：
 * - host：broker监听的主机名或IP地址
 * - port：监听的端口号
 * - listenerName：监听器名称，用于标识不同的监听配置
 * - securityProtocol：安全协议类型，如PLAINTEXT、SSL等
 */
case class EndPoint(host: String, port: Int, listenerName: ListenerName, securityProtocol: SecurityProtocol) {
  /**
   * 生成标准格式的连接字符串
   * 格式为：listenerName://host:port
   * 如果host为null，则只显示端口号
   * 
   * @return 格式化的连接字符串
   */
  def connectionString: String = {
    // 根据host是否为null构造主机端口部分
    val hostport =
      if (host == null)
        ":"+port
      else
        Utils.formatAddress(host, port)
    // 组合成最终的连接字符串
    listenerName.value + "://" + hostport
  }

  /**
   * 将Scala版本的EndPoint转换为Java版本的Endpoint
   * 用于在Scala和Java代码之间进行互操作
   * 
   * @return Java版本的Endpoint对象
   */
  def toJava: JEndpoint = {
    new JEndpoint(listenerName.value, securityProtocol, host, port)
  }
}
